src/dali/abstract_pair_converter_plugin.py

```diff
@@ -88,6 +93,45 @@ def get_or_set_vertex(self, name: str) -> Vertex[ValueType]:
         self.add_vertex(new_vertex)
         return new_vertex
 
+    def is_optimized(self, asset: str) -> bool:
+        return asset in self.__optimized_assets
+
+    # Optimized asset is the asset we are optimizing during this cloning.
+    # Optimized assets are tracked to prevent redundant cloning.
+    # Negative weights will get deleted.
+    def clone_with_optimization(self, optimized_asset: str, neighbor_weights: Dict[str, float]) -> "MappedGraph[ValueType]":
+        if optimized_asset in self.__optimized_assets:
+            raise RP2ValueError(
+                "You are trying to optimize an asset that has already been optimized."
+                "Use is_optimized() to check if an asset has already been optimized before cloning."
+            )
+
+        cloned_optimized_assets: Dict[str, None] = self.__optimized_assets
+        cloned_optimized_assets[optimized_asset] = None
+        cloned_mapped_graph: MappedGraph[ValueType] = MappedGraph(optimized_assets=cloned_optimized_assets)
+
+        for original_vertex in self.vertexes:
+            # Add existing neighbors
+            for neighbor in original_vertex.neighbors:
+                neighbor_weight: float
+                if original_vertex.name == optimized_asset:
+                    neighbor_weight = neighbor_weights.pop(neighbor.name, original_vertex.get_weight(neighbor))
+                else:
+                    neighbor_weight = original_vertex.get_weight(neighbor)
+
+                # Delete neighbor if negative weight
+                if neighbor_weight >= 0.0:
+                    cloned_mapped_graph.add_neighbor(original_vertex.name, neighbor.name, neighbor_weight)
+                else:
+                    cloned_mapped_graph.add_missing_vertex(original_vertex.name)
+
+            # Add new neighbors
+            for neighbor_name in neighbor_weights.keys():
+                if original_vertex.name == optimized_asset:
+                    cloned_mapped_graph.add_neighbor(original_vertex.name, neighbor_name, neighbor_weights[neighbor_name])
```

Comment:
Could the neighbors be already added? If not can you clarify in a comment why?

Reply:
The `add_neighbor()` function is overwritten to prevent adding a neighbor that has already been added.